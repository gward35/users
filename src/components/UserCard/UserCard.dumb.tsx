import {
  Avatar,
  Card,
  CardBody,
  ChakraProps,
  Flex,
  Spinner,
  Stack,
  Text,
} from '@chakra-ui/react'
import { ModifiedUsers } from '@/types/userTypes'
import { getTextColor } from '../ColorInput/logic/logic'

export interface DumbUserCardProps extends ChakraProps {
  cardColor: string
  modifiedUsers: ModifiedUsers[]
  isLargerThan768: boolean
  isLoading: boolean
  isValid: boolean
}

export const DumbUserCard = ({
  cardColor,
  modifiedUsers,
  isLargerThan768,
  isLoading,
  isValid,
}: DumbUserCardProps) => {
  const setTextColor = !isValid
    ? 'black'
    : getTextColor(cardColor)
    ? 'white'
    : 'black'
  return (
    <>
      <Flex
        justifyContent="center"
        alignItems="center"
        flexDirection={['column', 'row']}
        flexWrap="wrap"
        py={8}
        mx={4}
      >
        {modifiedUsers.map((u: ModifiedUsers, i: number) => (
          <Card
            key={u.id.value + i}
            alignItems="center"
            {...(isValid ? { bgColor: cardColor } : {})}
            border="1px solid"
            borderColor="lightGray.100"
            boxShadow="5px 5px 10px #D8DDE8"
            data-testid="userCard"
            mr={[0, 0, 4]}
            mb={4}
            padding={4}
            flex={isLargerThan768 ? '1 0 30%' : '1 0 80%'}
            sx={{
              [':nth-of-type(3n)']: { mr: 0 },
            }}
          >
            <Avatar
              borderRadius="100%"
              color="black"
              name={`${u.name.first ?? 'Unknown'} ${u.name.last ?? 'Unknown'}`}
              src={u.picture.large}
              srcSet={`${u.picture.thumbnail} 480w, ${u.picture.medium} 800w`}
              mb={4}
              height={20}
              width={20}
            />
            <Stack>
              <CardBody>
                <Text
                  as="p"
                  color={setTextColor}
                  data-testid="name"
                  fontSize="sm"
                >
                  {u.name.first ?? 'Unknown'} {u.name.last ?? 'Unknown'}
                </Text>
                <Text
                  as="p"
                  color={setTextColor}
                  data-testid="location"
                  fontSize="sm"
                >
                  {u.location.city ?? 'Unknown'},{' '}
                  {u.location.country ?? 'Unknown'}
                </Text>
                <Text
                  as="p"
                  color={setTextColor}
                  data-testid="email"
                  fontSize="sm"
                >
                  {u.email ?? 'Unknown'}
                </Text>
                <Text
                  as="p"
                  color={setTextColor}
                  data-testid="phoneNumber"
                  fontSize="sm"
                >
                  {u.cell ? u.cell : u.phone}
                </Text>
              </CardBody>
            </Stack>
          </Card>
        ))}
      </Flex>
      {isLoading && (
        <Flex justifyContent="center">
          <Spinner
            thickness="6px"
            speed="0.65s"
            emptyColor="lightGray.100"
            color="primary.100"
            size="xl"
            mb={8}
          />
        </Flex>
      )}
    </>
  )
}

export default DumbUserCard
