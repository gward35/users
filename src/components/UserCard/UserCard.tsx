import React, { useEffect, useState, useMemo } from 'react'
import DumbUserCard from './UserCard.dumb'
import { useUsers } from '@/api/userApi'
import { Results, ModifiedUsers } from '@/types/userTypes'
import Error from '../Error/Error'

export const modifyUsers = (users: ModifiedUsers[]) =>
  users.map(
    ({ name, cell, email, phone, picture, location, id }: ModifiedUsers) => ({
      name,
      cell,
      email,
      phone,
      picture,
      location,
      id,
    })
  )

export type UserCardProps = {
  cardColor: string
  isLargerThan768: boolean
  isValid: boolean
  page: number
}

export const UserCard = ({
  cardColor,
  isLargerThan768,
  isValid,
  page,
}: UserCardProps) => {
  const { data, error, isLoading } = useUsers(page)
  const results: Results[] = useMemo(() => {
    return data?.results ?? []
  }, [data])
  const modifiedUsers: ModifiedUsers[] = modifyUsers(results)
  const [$modifiedUsers, $setModifiedUsers] =
    useState<ModifiedUsers[]>(modifiedUsers)

  useEffect(() => {
    $setModifiedUsers(prevState => [...prevState, ...results])
  }, [results])

  if (error) return <Error />

  return (
    <DumbUserCard
      isLargerThan768={isLargerThan768}
      isLoading={isLoading}
      isValid={isValid}
      cardColor={cardColor}
      modifiedUsers={$modifiedUsers}
    />
  )
}

export default UserCard
