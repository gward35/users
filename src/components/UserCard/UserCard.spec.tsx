import { render } from '@testing-library/react'
import UserCard, { modifyUsers } from './UserCard'
import { API_RESULT } from '@/testing/constants'

describe('UserCard', () => {
  const PROPS = {
    cardColor: '',
    page: 1,
    isLargerThan768: true,
    isValid: true,
  }

  it('renders component and snaps', () => {
    const utils = render(<UserCard {...PROPS} />)
    expect(utils.container).toMatchSnapshot()
  })

  it('tests constant response', () => {
    expect(modifyUsers(API_RESULT)).toEqual([
      {
        cell: '987-654-3210',
        email: 'a@b.com',
        id: {
          name: 'C',
          value: 'D',
        },
        location: {
          city: 'City',
          coordinates: {
            latitude: '30.5',
            longitude: '95.5',
          },
          country: 'Country',
          postcode: 123456,
          state: 'State',
          street: {
            name: 'Main St',
            number: 123,
          },
          timezone: {
            description: 'Main St USA',
            offset: '+1:00',
          },
        },
        name: {
          first: 'A',
          last: 'B',
          title: 'Honorable',
        },
        phone: '012-345-6789',
        picture: {
          large: 'https://randomuser.me/api/portraits/women/88.jpg',
          medium: 'https://randomuser.me/api/portraits/med/women/88.jpg',
          thumbnail: 'https://randomuser.me/api/portrait',
        },
      },
    ])
  })
})
