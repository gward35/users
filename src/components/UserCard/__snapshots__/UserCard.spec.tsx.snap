// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard renders component and snaps 1`] = `
<div>
  <div
    class="css-ibdz6f"
  >
    <div
      class="css-0"
    >
      <h2
        class="chakra-text css-16ukv61"
        data-testid="header"
      >
        Oh No!
      </h2>
    </div>
    <p
      class="chakra-text css-10go69x"
      data-testid="errorMsg"
    >
      Sorry, an error has happened on our end! Please try refreshing the page. If the error persists please contact the
       
      <a
        class="chakra-link css-1wkcdqa"
        data-testid="errorLink"
        href="mailto:help@myusers.com"
        rel="noopener"
        target="_blank"
      >
        Help Center
      </a>
    </p>
  </div>
</div>
`;
